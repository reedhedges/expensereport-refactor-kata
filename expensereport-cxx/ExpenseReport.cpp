#include <chrono>
#include <iostream>
#include <iterator>
#include <list>

#include "ExpenseReport.h"

using namespace ExpenseReportKata;

void printReport(std::list<Expense> expenses)
{
    int total = 0;
    int mealExpenses = 0;

    auto now = std::chrono::system_clock::to_time_t(std::chrono::system_clock::now());
    std::cout << "Expenses " << ctime(&now) << '\n';

    for (std::list<Expense>::iterator expense = expenses.begin(); expense != expenses.end(); ++expense) {
        if (expense->type == BREAKFAST || expense->type == DINNER) {
            mealExpenses += expense->amount;
        }

        std::string expenseName = "";
        switch (expense->type) {
        case DINNER:
            expenseName = "Dinner";
            break;
        case BREAKFAST:
            expenseName = "Breakfast";
            break;
        case CAR_RENTAL:
            expenseName = "Car Rental";
            break;
        }

        std::string mealOverExpensesMarker = (expense->type == DINNER && expense->amount > 5000) || (expense->type == BREAKFAST && expense->amount > 1000) ? "X" : " ";

        std::cout << expenseName << '\t' << expense->amount << '\t' << mealOverExpensesMarker << '\n';

        total += expense->amount;
    }

    std::cout << "Meal expenses: " << mealExpenses << '\n';
    std::cout << "Total expenses: " << total << '\n';
}
