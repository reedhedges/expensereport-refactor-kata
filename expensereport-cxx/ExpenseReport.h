#include <chrono>
#include <iostream>
#include <iterator>
#include <list>

using namespace std;

enum Type
{
    BREAKFAST, DINNER, CAR_RENTAL
};

class Expense
{
    public:
    Type type;
    int amount;
};

void printReport(list<Expense> expenses)
{
    int total = 0;
    int mealExpenses = 0;

    auto now = chrono::system_clock::to_time_t(chrono::system_clock::now());
    cout << "Expenses " << ctime(&now) << '\n';

    for (list<Expense>::iterator expense = expenses.begin(); expense != expenses.end(); ++expense) {
        if (expense->type == BREAKFAST || expense->type == DINNER) {
            mealExpenses += expense->amount;
        }

        string expenseName = "";
        switch (expense->type) {
        case DINNER:
            expenseName = "Dinner";
            break;
        case BREAKFAST:
            expenseName = "Breakfast";
            break;
        case CAR_RENTAL:
            expenseName = "Car Rental";
            break;
        }

        string mealOverExpensesMarker = (expense->type == DINNER && expense->amount > 5000) || (expense->type == BREAKFAST && expense->amount > 1000) ? "X" : " ";

        cout << expenseName << '\t' << expense->amount << '\t' << mealOverExpensesMarker << '\n';

        total += expense->amount;
    }

    cout << "Meal expenses: " << mealExpenses << '\n';
    cout << "Total expenses: " << total << '\n';
}
