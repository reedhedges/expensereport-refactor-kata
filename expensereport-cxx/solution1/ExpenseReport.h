#pragma once

#include <chrono>
#include <list>
#include <optional>
#include <climits>
#include <string_view>
#include <memory>

#define KATA_SOLUTION_1 1
namespace ExpenseReportKata {

class Expense
{
    unsigned int amount_;
public:
    Expense(unsigned int amount) : amount_(amount) {}
    virtual ~Expense() = default;
    unsigned int amount() const { return amount_; }
    virtual std::string_view name() const { return "Unknown"; }
    virtual unsigned int limit() const { return UINT_MAX; }
    bool aboveLimit() const { return amount() > limit(); }
    enum Category { MEAL, NONMEAL };
    virtual Category category() const { return NONMEAL; }
};

// todo these could be moved to another header file and/or namespace
// todo many of these could be made static?
class BreakfastExpense : public virtual Expense
{
public:
    BreakfastExpense(unsigned int amount) : Expense(amount) {}
    virtual ~BreakfastExpense() = default;
    virtual std::string_view name() const override { return "Breakfast"; }
    virtual unsigned int limit() const override { return 1000; }
    virtual Category category() const override { return MEAL; }
};
class DinnerExpense : public virtual Expense
{
public:
    DinnerExpense(unsigned int amount) : Expense(amount) {}
    virtual ~DinnerExpense() = default;
    virtual std::string_view name() const override  { return "Dinner"; }
    virtual unsigned int limit() const override { return 5000; }
    virtual Category category() const override { return MEAL; }
};
class LunchExpense : public virtual Expense
{    
public:
    LunchExpense(unsigned int amount) : Expense(amount) {}
    virtual ~LunchExpense() = default;
    virtual std::string_view name() const override { return "Lunch"; }
    virtual unsigned int limit() const  override { return 2000; }
    virtual Category category() const override { return MEAL; }
};
class CarRentalExpense : public virtual Expense
{
public:
    CarRentalExpense(unsigned int amount) : Expense(amount) {}
    virtual ~CarRentalExpense() = default;
    virtual std::string_view name() const override { return "Car Rental"; }
};

using report_time_t = std::chrono::time_point<std::chrono::system_clock>;
using expense_list_t = std::list<std::unique_ptr<Expense>>;
void printReport(const expense_list_t& expenses, std::optional<report_time_t> time_opt = std::nullopt);

} // namespace ExpenseReportKata