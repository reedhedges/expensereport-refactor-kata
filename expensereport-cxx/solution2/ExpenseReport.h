#pragma once

#include <chrono>
#include <iterator>
#include <optional>
#include <array>
#include <climits>
#include <string>
#include <string_view>
#include <algorithm>
#include <numeric>

#define KATA_SOLUTION_2

namespace ExpenseReportKata {

enum Category {
    MEAL,
    NONMEAL
};

struct Type
{    
    std::string_view name; // note a Type struct is expected to be constructed with a static string literal for name
    unsigned int limit;
    Category category;
};

enum TypeIndex {
    BREAKFAST = 0,
    DINNER = 1,
    LUNCH = 2,
    CAR_RENTAL = 3,
    MAX_TYPE_INDEX = 3
};

namespace Private {
    // note: order in types array must match TypeIndex
    static constexpr std::array<Type, MAX_TYPE_INDEX+1> g_types {
        Type{
            .name = "Breakfast",
            .limit = 1000,
            .category = MEAL
        },
        Type{
            .name = "Dinner",
            .limit = 5000,
            .category = MEAL
        },
        Type{
            .name = "Lunch",
            .limit = 2000,
            .category = MEAL
        },
        Type{
            .name = "Car Rental",
            .limit = UINT_MAX,
            .category = NONMEAL
        }
    };
}

static constexpr const Type& getType(TypeIndex t) {
    return Private::g_types[t];
}

class Expense
{
private:
    const Type& type_;   
    unsigned int amount_;

public:
    constexpr Expense(TypeIndex t, unsigned int amt) : type_(getType(t)), amount_(amt) {}
    // default move and copy constructors and operators should work
    constexpr unsigned int amount() const { return amount_; }
    constexpr bool aboveLimit() const { return amount_ > type_.limit; }
    constexpr std::string_view name() const { return type_.name; }
    constexpr Category category() const { return type_.category; }
};

using report_time_t = std::chrono::time_point<std::chrono::system_clock>;

//bool printReport(const std::list<Expense>& expenses, std::optional<report_time_t> time_opt = std::nullopt);
template <typename It>
bool printReport(const It begin, const It end, std::optional<report_time_t> time_opt = std::nullopt)
{
    time_t now = 0;
    if(time_opt) now = std::chrono::system_clock::to_time_t(*time_opt);

    std::cout << "Expenses " << ctime(&now) << '\n';

    unsigned long total = 0;
    unsigned long mealExpenses = 0; // todo this could be abstracted to generic map of category -> total, with category names in static list/map, similar to Types. 

    //for (const auto& expense : expenses) {
    for(auto i = begin; i != end; ++i)
    {
        // todo guard against overflowing total or mealExpenses
        const Expense& expense = *i;
        if (expense.category() == MEAL) {
            mealExpenses += expense.amount();
        }
        total += expense.amount();
        std::cout << expense.name() << '\t' << expense.amount() << '\t' << (expense.aboveLimit() ? 'X' : ' ') << '\n';
    }

    std::cout << "Meal expenses: " << mealExpenses << '\n';
    std::cout << "Total expenses: " << total << '\n';

    return true;
}

template <typename T>
bool printReport(const T& cont, std::optional<report_time_t> time_opt = std::nullopt) {
    return printReport(std::cbegin(cont), std::cend(cont), time_opt);
}

template<size_t N>
constexpr unsigned long getTotal(const std::array<Expense, N>& expenses)
{
   /*  std::for_each(expenses.cbegin(), expenses.cend(), [&r](const auto& e){
        // todo guard against overflowing total or mealExpenses
        if (e.category() == MEAL) {
            r.mealExpenses += e.amount();
        }
        r.total += e.amount();
        if(e.aboveLimit())
            r.numOverLimit += 1;
    }); */
    return std::accumulate(expenses.cbegin(), expenses.cend(), 0UL, [](const auto sum, const Expense& next)
    {
        return sum + next.amount();
    });
}

template<size_t N>
constexpr unsigned long getMealsTotal(const std::array<Expense, N>& expenses)
{
    return std::accumulate(expenses.cbegin(), expenses.cend(), 0UL, [](const auto sum, const Expense& next)
    {
        if(next.category() == MEAL) return sum + next.amount();
        return sum;
    }
    );
}

template<size_t N>
constexpr size_t getNumAboveLimit(const std::array<Expense, N>& expenses)
{
    return std::count_if(expenses.cbegin(), expenses.cend(), [](const Expense& next)
    {
        return next.aboveLimit();
    });
}

} // namespace ExpenseReportKata