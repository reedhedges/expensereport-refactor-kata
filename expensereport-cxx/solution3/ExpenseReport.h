#pragma once

#include <chrono>
#include <iostream>
#include <iterator>
#include <optional>
#include <array>
#include <string_view>
#include <climits>

#define KATA_SOLUTION_3

namespace ExpenseReportKata {

enum Type
{
    BREAKFAST = 0, 
    LUNCH = 1, 
    DINNER = 2, 
    CAR_RENTAL = 3,
    MAX_TYPE_INDEX = 3
};

struct Expense
{
    Type type;
    unsigned int amount;
};

enum Category 
{
    MEAL, NONMEAL
};

namespace priv {
    // these could also be private members of Expense.
    // note order of these arrays must match order in Type enum
    constexpr static std::array<unsigned int, MAX_TYPE_INDEX+1> g_limits = { 
        1000,
        2000,
        5000,
        UINT_MAX
    };
    constexpr static std::array<std::string_view, MAX_TYPE_INDEX+1> g_names = {
        std::string_view{"Breakfast"},
        std::string_view{"Lunch"},
        std::string_view{"Dinner"},
        std::string_view{"Car Rental"}
    };
    constexpr static std::array<Category, MAX_TYPE_INDEX+1> g_categories = {
        MEAL,
        MEAL,
        MEAL,
        NONMEAL
    };
}

// note these could also be member functions of Expense:

unsigned int limit(const Expense& expense) {
    return priv::g_limits[expense.type]; 
}

bool aboveLimit(const Expense& expense) {
    return expense.amount > limit(expense);
}

bool category(const Expense& expense) {
    return priv::g_categories[expense.type];
}

std::string_view name(const Expense& expense) {
    return priv::g_names[expense.type];
}

unsigned int amount(const Expense& expense) {
    return expense.amount;
}

using report_time_t = std::chrono::time_point<std::chrono::system_clock>;

template<typename It>
void printReport(const It begin, const It end, std::optional<report_time_t> time_opt = std::nullopt)
{
    unsigned int total = 0;
    unsigned int mealExpenses = 0;

    const auto now = std::chrono::system_clock::to_time_t(time_opt ? *time_opt : std::chrono::system_clock::now());
    std::cout << "Expenses " << ctime(&now) << '\n';

    for (auto i = begin; i != end; ++i)
    {
        const auto& e = *i;
        const auto amt = amount(e);
        if (category(e) == MEAL) { // could also make generic map of category->categoryTotal.
            mealExpenses += amt;
        }
        total += amt;
        std::cout << name(e) << '\t' << amt << '\t' << (aboveLimit(e) ? 'X' : ' ') << '\n';
    }

    std::cout << "Meal expenses: " << mealExpenses << '\n';
    std::cout << "Total expenses: " << total << '\n';
}

template <typename T>
void printReport(const T& expenses, std::optional<report_time_t> time_opt = std::nullopt)
{
    printReport(cbegin(expenses), cend(expenses), time_opt);
}

} // namespace ExpenseReportKata