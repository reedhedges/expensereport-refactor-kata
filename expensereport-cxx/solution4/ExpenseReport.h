#pragma once

#include <chrono>
#include <iostream>
#include <iterator>
#include <optional>
#include <string_view>
#include <array>
#include <climits>

#define KATA_SOLUTION_4
#define KATA_SOLUTION_CONSTEXPR 

namespace ExpenseReportKata {

enum Type
{
    BREAKFAST, 
    LUNCH, 
    DINNER, 
    CAR_RENTAL
};

struct Expense
{
    Type type;
    unsigned int amount;
};

enum Category 
{
    MEAL, NONMEAL
};

// note these could also be member functions of Expense:

constexpr unsigned int limit(const Expense& expense) {
    switch (expense.type) {
        case BREAKFAST:     return 1000;
        case LUNCH:         return 2000;
        case DINNER:        return 5000;
        default:            return UINT_MAX;
    }
}

constexpr bool aboveLimit(const Expense& expense) {
    return expense.amount > limit(expense);
}

constexpr bool category(const Expense& expense) {
    switch (expense.type) {
        case BREAKFAST:
        case LUNCH:
        case DINNER:
            return MEAL;
        default:    
            return NONMEAL;
    }
}

constexpr std::string_view name(const Expense& expense) {
    switch (expense.type) {
        case BREAKFAST:     return "Breakfast";
        case LUNCH:         return "Lunch";
        case DINNER:        return "Dinner";
        case CAR_RENTAL:    return "Car Rental";
    }
    return "Unknown expense type";
}

unsigned int amount(const Expense& expense) {
    return expense.amount;
}

using report_time_t = std::chrono::time_point<std::chrono::system_clock>;

// expenses must be an iterable container of Expense objects.
template<typename It>
void printReport(const It begin, const It end, std::optional<report_time_t> time_opt = std::nullopt)
{
    unsigned int total = 0;
    unsigned int mealExpenses = 0;

    const auto now = std::chrono::system_clock::to_time_t(time_opt ? *time_opt : std::chrono::system_clock::now());
    std::cout << "Expenses " << ctime(&now) << '\n';

    for (auto i = begin; i != end; ++i)
    {
        const auto& e = *i;
        if (category(e) == MEAL) { // could also make generic map of category->categoryTotal.
            mealExpenses += amount(e);
        }
        total += amount(e);
        std::cout << name(e) << '\t' << amount(e) << '\t' << (aboveLimit(e) ? 'X' : ' ') << '\n';
    }

    std::cout << "Meal expenses: " << mealExpenses << '\n';
    std::cout << "Total expenses: " << total << '\n';
}

template <typename T>
void printReport(const T& expenses, std::optional<report_time_t> time_opt = std::nullopt)
{
    printReport(cbegin(expenses), cend(expenses), time_opt);
}



// todo allow other containers other than array: (e.g. implement with iterators, then add wrappers for array or non-array containers)

template<size_t N>
constexpr unsigned long getTotal(const std::array<Expense, N>& expenses)
{
    return std::accumulate(expenses.cbegin(), expenses.cend(), 0UL, [](const auto sum, const Expense& next)
    {
        return sum + next.amount;
    });
}

template<size_t N>
constexpr unsigned long getCategoryTotal(const std::array<Expense, N>& expenses, Category cat)
{
    return std::accumulate(expenses.cbegin(), expenses.cend(), 0UL, [cat](const auto sum, const Expense& next)
    {
        if(category(next) == cat) return sum + next.amount;
        return sum;
    }
    );
}

template<size_t N>
constexpr unsigned long getMealsTotal(const std::array<Expense, N>& expenses)
{
    return getCategoryTotal(expenses, MEAL);
}

template<size_t N>
constexpr size_t getNumAboveLimit(const std::array<Expense, N>& expenses)
{
    return std::count_if(expenses.cbegin(), expenses.cend(), [](const Expense& next)
    {
        return aboveLimit(next);
    });
}

} // namespace ExpenseReportKata