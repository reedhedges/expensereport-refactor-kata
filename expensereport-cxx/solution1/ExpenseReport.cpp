#include <chrono>
#include <iostream>
#include <iterator>
#include <list>
#include <optional>
#include <cassert>

#include "ExpenseReport.h"

using namespace ExpenseReportKata;

void ExpenseReportKata::printReport(const expense_list_t& expenses, std::optional<report_time_t> time_opt)
{
    const auto now = std::chrono::system_clock::to_time_t(time_opt ? *time_opt : std::chrono::system_clock::now());
    std::cout << "Expenses " << ctime(&now) << '\n';

    unsigned long total = 0;
    unsigned long mealExpenses = 0; // todo this could be a generic map of category -> total, with category names in category subclasses, similar to expense types.
    for (auto i = expenses.cbegin(); i != expenses.cend(); ++i) {
        const Expense& expense = **i; // i is iterator to unique_ptr
        // todo guard against overflow in total and mealExpenses
        if (expense.category() == Expense::MEAL) {
            mealExpenses += expense.amount();
        }
        total += expense.amount();
        std::cout << expense.name() << '\t' << expense.amount() << '\t' << (expense.aboveLimit() ? 'X' : ' ') << '\n';
    }

    std::cout << "Meal expenses: " << mealExpenses << '\n';
    std::cout << "Total expenses: " << total << '\n';
}
