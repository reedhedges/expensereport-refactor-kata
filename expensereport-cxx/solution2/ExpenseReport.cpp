#include <chrono>
#include <iostream>
#include <iterator>
#include <list>
#include <optional>

#include "ExpenseReport.h"

using namespace ExpenseReportKata;


bool ExpenseReportKata::printReport(const std::list<Expense>& expenses, std::optional<report_time_t> time_opt)
{
    time_t now = 0;
    if(time_opt) now = std::chrono::system_clock::to_time_t(*time_opt);

    std::cout << "Expenses " << ctime(&now) << '\n';

    unsigned long total = 0;
    unsigned long mealExpenses = 0;

    for (const auto& expense : expenses) {
        // todo guard against overflowing total or mealExpenses
        if (expense.category() == MEAL) {
            mealExpenses += expense.amount();
        }
        total += expense.amount();
        std::cout << expense.name() << '\t' << expense.amount() << '\t' << (expense.aboveLimit() ? 'X' : ' ') << '\n';
    }

    std::cout << "Meal expenses: " << mealExpenses << '\n';
    std::cout << "Total expenses: " << total << '\n';

    return true;
}
