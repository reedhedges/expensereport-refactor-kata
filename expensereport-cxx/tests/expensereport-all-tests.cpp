#define APPROVAL_TESTS_HIDE_DEPRECATED_CODE 1

// defines main function:
#define APPROVALS_CATCH

#include "ApprovalTests.hpp"
#include "catch.hpp"

using namespace ApprovalTests;

#include "ExpenseReport.h"
#include <vector>
#include <list>
#include <iostream>
#include <sstream>
#include <string_view>
#include <memory>

// During this object's lifetime, all output to std::cout will instead be stored in a stringstream within this object. Use str() to get this string.
// Upon destruction, std::cout's output buffer is restored to its previous state (i.e. probably program stdout)
class CaptureCout {
    std::streambuf *orig_cout_buf;
    std::ostringstream outsstream;
public:
    CaptureCout() : orig_cout_buf(std::cout.rdbuf()) {
        std::cout.rdbuf(outsstream.rdbuf());
    }
    ~CaptureCout() {
        std::cout.rdbuf(orig_cout_buf);
    }
    std::string str() { return outsstream.str(); }
    // Copying and moving is disabled for this class
    // TODO implement these?
    CaptureCout(const CaptureCout& other) = delete;
    CaptureCout(CaptureCout&& old) = delete;
    CaptureCout& operator=(const CaptureCout& other) = delete;
    CaptureCout& operator=(CaptureCout&& old) = delete;
};

const auto system_clock_zero = std::chrono::time_point<std::chrono::system_clock>();

using namespace ExpenseReportKata;


#ifdef KATA_SOLUTION_1
TEST_CASE("Empty")
{
    // ApprovalTests will use the contents of the file expensereport-all-tests.Empty.approved.txt as the approved text compared against the argument to verify()
    CaptureCout captured;
    // Verify output of empty list:
    ExpenseReportKata::expense_list_t emptylist;
    printReport(emptylist, system_clock_zero); // use constant time value
    Approvals::verify(captured.str());
}
#else
TEST_CASE("Empty")
{
    CaptureCout cap;
    std::list<Expense> emptylist;
    //printReport(emptylist, system_clock_zero);
    printReport(emptylist.begin(), emptylist.end(), system_clock_zero);
    Approvals::verify(cap.str());
}

TEST_CASE("EmptyVector")
{
    CaptureCout cap;
    std::vector<Expense> emptylist;
    //printReport(emptylist, system_clock_zero);
    printReport(emptylist.begin(), emptylist.end(), system_clock_zero);
    Approvals::verify(cap.str());
}
#endif

#ifdef KATA_SOLUTION_1
TEST_CASE("Several_AntiRegression")
{    
    // This report output was approved from the original version, before refactoring.
    // This test will break if formatting of report changes, or any limits of the types changes.
    ExpenseReportKata::expense_list_t e;
    e.push_back(std::make_unique<BreakfastExpense>(1000));
    e.push_back(std::make_unique<BreakfastExpense>(1001));
    e.push_back(std::make_unique<DinnerExpense>(5000));
    e.push_back(std::make_unique<DinnerExpense>(5001));
    e.push_back(std::make_unique<BreakfastExpense>(0));
    e.push_back(std::make_unique<DinnerExpense>(0));
    e.push_back(std::make_unique<CarRentalExpense>(1000));
    e.push_back(std::make_unique<CarRentalExpense>(999999));
    e.push_back(std::make_unique<CarRentalExpense>(0));
    CaptureCout capt;
    ExpenseReportKata::printReport(e, system_clock_zero);
    Approvals::verify(capt.str());
}
#else
TEST_CASE("Several_AntiRegression")
{    
    // This report output was approved from the original version, before refactoring.
    // This test will break if formatting of report changes, or any limits of the types changes.
    std::list<Expense> e;
    e.emplace_back(BREAKFAST, 1000);
    e.emplace_back(BREAKFAST, 1001);
    e.emplace_back(DINNER, 5000);
    e.emplace_back(DINNER, 5001);
    e.emplace_back(BREAKFAST, 0);
    e.emplace_back(DINNER, 0);
    e.emplace_back(CAR_RENTAL, 1000);
    e.emplace_back(CAR_RENTAL, 999999);
    e.emplace_back(CAR_RENTAL, 0);
    CaptureCout capt;
//    ExpenseReportKata::printReport(e, system_clock_zero);
    ExpenseReportKata::printReport(e.cbegin(), e.cend(), system_clock_zero);
    Approvals::verify(capt.str());
}
#endif

TEST_CASE("ExpenseSize")
{
    if(sizeof(ExpenseReportKata::Expense) > 8) {
        WARN("sizeof(Expense) > 8: " << sizeof(Expense));
    }
}

#ifdef KATA_SOLUTION_1
TEST_CASE("DerivedExpenseSize")
{
    if(sizeof(BreakfastExpense) > 8) {
        WARN("sizeof(BreakfastExpense) > 8: " << sizeof(BreakfastExpense));
    }
}
#endif


#ifdef KATA_SOLUTION_1

TEST_CASE("Several_WithFixedAmounts") 
{
    // This test will break if formatting of report changes, or any limits of the types changes.
    ExpenseReportKata::expense_list_t e;
    e.push_back(std::make_unique<BreakfastExpense>(1000));
    e.push_back(std::make_unique<BreakfastExpense>(1001));
    e.push_back(std::make_unique<DinnerExpense>(5000));
    e.push_back(std::make_unique<DinnerExpense>(5001));
    e.push_back(std::make_unique<LunchExpense>(2000));
    e.push_back(std::make_unique<LunchExpense>(2001));

    CaptureCout capt;
    ExpenseReportKata::printReport(e, system_clock_zero);
    Approvals::verify(capt.str());
}

TEST_CASE("Several_WithRelAmounts") 
{
    // This test will break if report changes, but not if limit values of types change, since we use values based on whatever the type's limit value is

    BreakfastExpense bkfst(0);
    DinnerExpense dinner(0);
    LunchExpense lunch(0);

    ExpenseReportKata::expense_list_t e;
    e.push_back(std::make_unique<BreakfastExpense>(bkfst.limit()));
    e.push_back(std::make_unique<BreakfastExpense>(bkfst.limit() + 1));
    e.push_back(std::make_unique<DinnerExpense>(dinner.limit()));
    e.push_back(std::make_unique<DinnerExpense>(dinner.limit() + 1));
    e.push_back(std::make_unique<LunchExpense>(lunch.limit()));
    e.push_back(std::make_unique<LunchExpense>(lunch.limit() + 1));

    CaptureCout capt;
    ExpenseReportKata::printReport(e, system_clock_zero);
    Approvals::verify(capt.str());
}

#else

TEST_CASE("Several_WithFixedAmounts") 
{
    // This test will break if formatting of report changes, or any limits of the types changes.
    std::list<Expense> e;
    e.emplace_back(BREAKFAST, 1000);
    e.emplace_back(BREAKFAST, 1001);
    e.emplace_back(DINNER, 5000);
    e.emplace_back(DINNER, 5001);
    e.emplace_back(LUNCH, 2000);
    e.emplace_back(LUNCH, 2001);

    CaptureCout capt;
    ExpenseReportKata::printReport(e, system_clock_zero);
    Approvals::verify(capt.str());
}

TEST_CASE("Several_WithFixedAmounts_Vector") 
{
    // This test will break if formatting of report changes, or any limits of the types changes.
    std::vector<Expense> e;
    e.emplace_back(BREAKFAST, 1000);
    e.emplace_back(BREAKFAST, 1001);
    e.emplace_back(DINNER, 5000);
    e.emplace_back(DINNER, 5001);
    e.emplace_back(LUNCH, 2000);
    e.emplace_back(LUNCH, 2001);

    CaptureCout capt;
    ExpenseReportKata::printReport(e, system_clock_zero);
    Approvals::verify(capt.str());
}

TEST_CASE("Several_WithRelAmounts") 
{
    // This test will break if report changes, but not if limit values of types change, since we use values based on whatever the type's limit value is

    const ExpenseReportKata::Type& bkfst = getType(BREAKFAST);
    const ExpenseReportKata::Type& dinner = getType(DINNER);
    const ExpenseReportKata::Type& lunch = getType(LUNCH);

    CHECK(bkfst.category == MEAL);
    CHECK(dinner.category == MEAL);
    CHECK(lunch.category == MEAL);

    std::list<Expense> e;
    e.emplace_back(BREAKFAST, bkfst.limit);
    e.emplace_back(BREAKFAST, bkfst.limit + 1);
    e.emplace_back(DINNER, dinner.limit);
    e.emplace_back(DINNER, dinner.limit + 1);
    e.emplace_back(LUNCH, lunch.limit);
    e.emplace_back(LUNCH, lunch.limit + 1);

    CaptureCout capt;
    ExpenseReportKata::printReport(e, system_clock_zero);
    Approvals::verify(capt.str());
}

#endif


#ifdef KATA_SOLUTION_2
#include <array>
TEST_CASE("Constexpr")
{
    constexpr std::array<Expense, 3> expenses {
        Expense {BREAKFAST, 800},
        Expense {LUNCH, 1200},
        Expense {DINNER, 5200} // should be above limit
    };
    static_assert(getTotal(expenses) == 800+1200+5200);
    STATIC_REQUIRE(getTotal(expenses) == 800+1200+5200);
    STATIC_REQUIRE(getTotal(expenses) == getMealsTotal(expenses)); // all are meals
    STATIC_REQUIRE(getNumAboveLimit(expenses) == 1);
}
#endif