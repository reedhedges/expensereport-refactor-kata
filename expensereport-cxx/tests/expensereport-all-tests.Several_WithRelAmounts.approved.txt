Expenses Wed Dec 31 19:00:00 1969

Breakfast	1000	 
Breakfast	1001	X
Dinner	5000	 
Dinner	5001	X
Lunch	2000	 
Lunch	2001	X
Meal expenses: 16003
Total expenses: 16003

